## Chapter 2 - Exercise 1: Facebook Advertising Analysis - Part 1

####  Facebook Advertising (Facebook Ads) là tên dịch vụ quảng cáo Facebook, cho phép hiển thị thông tin quảng cáo của các tổ chức/ cá nhân trên mạng xã hội Facebook. Quảng cáo trên mạng Facebook được phân phối tự động dựa trên rất nhiều các tiêu chí về giới tính, độ tuổi, khu vực địa lý, đặc điểm hành vi người dùng phù hợp với khách hàng mục tiêu của doanh nghiệp.
- CPC (Cost per Click) là hình thức quảng cáo trả phí được thanh toán dựa vào số click (bấm) vào quảng cáo. Ví dụ: khi người dùng trên Facebook bấm vào quảng cáo của bạn thì bạn sẽ mất một khoản tiền cho mỗi click. CPC tốt cho Lead Generation ( tìm kiếm và dẫn dụ người dùng). Ưu điểm lớn nhất của CPC đó chính là việc bạn sẽ tối ưu được ngân sách quảng cáo. Cụ thể trong trường hợp quảng cáo hiển thị đến những đối tượng người dùng không có nhu cầu với sản phẩm, dịch vụ, họ không nhấp vào quảng cáo thì doanh nghiệp bạn sẽ không mất bất kì khoản phí nào cả. 
### $ CPC = \frac{Money \ Spent}{Clicks} $
- CPM (Cost per mille hay “cost per thousand” là hình thức thanh toán dựa vào số lượt hiển thị quảng cáo, không cần biết user có tương tác hay không. Ví dụ: với cứ 1000 lượt hiển thị quảng cáo doanh nghiệp sẽ mất một khoản tiền cho 1000 lượt hiển thị đó. Hình thức quảng cáo này phù hợp nhất với các doanh nghiệp đang có nhu cầu phát triển thương hiệu nhiều hơn là bán hàng. Quảng cáo CPM cũng giúp đảm bảo rằng quảng cáo của bạn sẽ được tiếp cận đến tối đa người dùng trong tập đối tượng mục tiêu một cách liên tục. Nhược điểm đối với cách tính phí theo lượt hiển thị này đó chính là doanh nghiệp sẽ khó lòng nhắm quảng cáo đến đúng những đối tượng khách hàng tiềm năng. 
### $ CPM = (\frac{Money \ Spent}{Impressions}) \times 1000 $
![CPI](Data/CPC_CPI.png)

#### Cho các tập tin clicks.txt (chứa clicks), spent.txt (chứa money spent), và impressions.txt (chứa impressions).
#### Học viên thực hiện các yêu cầu sau, và đối chiếu với kết quả cho trước:

In [1]:
import numpy as np
import warnings
warnings.filterwarnings('ignore')

In [63]:
# Câu 1: Đọc dữ liệu từ 3 tập tin trên, và lưu vào 3 numpy array: clicks, spent và impressions chứa 3 thông tin (mỗi tập tin chứa một thông tin) của các mẫu.
# Gợi ý: (dùng np.loadtxt('ten_tap_tin', dtype=kieu_du_lieu))
clicks = np.loadtxt('clicks.txt', dtype=int)
spent = np.loadtxt('spent.txt', dtype=float)
impressions = np.loadtxt('impressions.txt', dtype=float)


In [3]:
# Câu 2: Cho biết mỗi numpy array trên chứa bao nhiêu mẫu. 
print('So mau cua clicks: %d mau'%clicks.size)
print('So mau cua spent: %d mau'%spent.size)
print('So mau cua impressions: %d mau'%impressions.size)

So mau cua clicks: 1143 mau
So mau cua spent: 1143 mau
So mau cua impressions: 1143 mau


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>Clicks chứa  1143 mẫu
Spent chứa  1143 mẫu
Impressions chứa  1143 mẫu
</pre>
  
</details>

In [5]:
# In 5 mẫu đầu tiên của từng array.
print('5 mẫu đầu tiên của clicks: ',clicks[0:5])
print('5 mẫu đầu tiên của spent: ',spent[0:5])
print('5 mẫu đầu tiên của impressions: ',impressions[0:5])

5 mẫu đầu tiên của clicks:  [1 2 0 1 1]
5 mẫu đầu tiên của spent:  [1.43 1.82 0.   1.25 1.29]
5 mẫu đầu tiên của impressions:  [ 7350. 17861.   693.  4259.  4133.]


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>5 mẫu đầu tiên của clicks:  [1 2 0 1 1]
5 mẫu đầu tiên của spent:  [1.43 1.82 0.   1.25 1.29]
5 mẫu đầu tiên của impressions:  [ 7350 17861   693  4259  4133]
</pre>
  
</details>

In [9]:
# Câu 3: Cho biết giá trị lần lượt của click, spent và impression ở vị trí 10, 100, 1000 của các numpy array trên.
print('Giá trị của click ở vị trí 10:%d vị trí 100:%d vị trí 1000:%d'%(clicks[10],clicks[100],clicks[1000]))
print('Giá trị của spent ở vị trí 10:%.2f vị trí 100:%.2f vị trí 1000:%.2f'%(spent[10],spent[100],spent[1000]))
print('Giá trị của impression ở vị trí 10:%.2f vị trí 100:%.2f vị trí 1000:%.2f'%(impressions[10],impressions[100],impressions[1000]))


Giá trị của click ở vị trí 10:0 vị trí 100:2 vị trí 1000:164
Giá trị của spent ở vị trí 10:0.00 vị trí 100:2.35 vị trí 1000:226.03
Giá trị của impression ở vị trí 10:1224.00 vị trí 100:6907.00 vị trí 1000:699314.00


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>Giá trị lần lượt của click ở vị trí 10, 100, 1000:  [  0   2 164]
Giá trị lần lượt của spent ở vị trí 10, 100, 1000:  [  0.     2.35 226.03]
Giá trị lần lượt của impression ở vị trí 10, 100, 1000:  [  1224   6907 699314]
</pre>
  
</details>

In [13]:
# Câu 4: Cho biết giá trị 5 mẫu cuối cùng của các numpy array trên
print('Giá trị 5 mẫu cuối cùng của clicks:',clicks[-5:])
print('Giá trị 5 mẫu cuối cùng của spent:',spent[-5:])
print('Giá trị 5 mẫu cuối cùng của impressions:',impressions[-5:])

Giá trị 5 mẫu cuối cùng của clicks: [252 120  28 135 114]
Giá trị 5 mẫu cuối cùng của spent: [358.19 173.88  40.29 198.71 165.61]
Giá trị 5 mẫu cuối cùng của impressions: [1129773.  637549.  151531.  790253.  513161.]


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>5 mẫu cuối cùng của clicks:  [252 120  28 135 114]
5 mẫu cuối cùng của spent:  [358.19 173.88  40.29 198.71 165.61]
5 mẫu cuối cùng của impressions:  [1129773  637549  151531  790253  513161]
</pre>
  
</details>

In [14]:
# Câu 5: Tính CPC của từng mẫu theo công thức trên.
cpc = spent/clicks
# Hiển thị giá trị CPC của 5 mẫu đầu
print('Giá trị CPC của 5 mẫu đầu:',cpc[:5])

Giá trị CPC của 5 mẫu đầu: [1.43 0.91  nan 1.25 1.29]


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>array([1.43, 0.91,  nan, 1.25, 1.29])</pre>
  
</details>

In [15]:
# Câu 6: Tính CPM cho từng mẫu theo công thức trên.
cpm = (spent/impressions) * 1000
# Hiển thị giá trị CPM của 5 mẫu đầu
print('Giá trị CPM của 5 mẫu đầu:',cpm[:5])

Giá trị CPM của 5 mẫu đầu: [0.19455782 0.10189799 0.         0.29349613 0.31212195]


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>array([0.19455782, 0.10189799, 0.        , 0.29349613, 0.31212195])</pre>
  
</details>

In [ ]:
# Câu 7: Tìm giá trị clicks lớn nhất, từ đó cho biết mức spent tương ứng.

In [20]:
max_click = np.max(clicks)
print('Giá trị click lớn nhất',max_click)

Giá trị click lớn nhất 421


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>Giá trị clicks lớn nhất:  421
</pre>
  
</details>

In [21]:
max_click_index = np.where(clicks == max_click)
max_click_index

(array([860]),)

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
    <pre>Vị trí clicks lớn nhất: 
</pre>
  <pre>(array([860], dtype=int64),)</pre>
  
</details>

In [22]:
print('Mức spent tương ứng',spent[max_click_index])

Mức spent tương ứng [612.3]


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>Giá trị spent tương ứng là: 
</pre>
<pre>array([612.3])</pre>
  
</details>

In [ ]:
# Câu 8: Tìm giá trị/ vị trí impressions nhỏ nhất và lớn nhất. Từ đó, cho biết spent tương ứng.

In [26]:
min_impression = np.min(impressions)
min_impression

87.0

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>Giá trị impressions nhỏ nhất:  87
</pre>
  
</details>

In [30]:
min_impression_index = np.where(impressions == min_impression)
min_impression_index

(array([500]),)

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>Vị trí impressions nhỏ nhất: 
</pre>
  <pre>(array([500], dtype=int64),)</pre>
</details>

In [32]:
print('Spent tương ứng với impression nhỏ nhất:',spent[min_impression_index])

Spent tương ứng với impression nhỏ nhất: [0.]


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>Giá trị spent tương ứng là: 
</pre>
  <pre>array([0.])</pre>
</details>

In [33]:
max_impression = np.max(impressions)
max_impression

3052003.0

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>Giá trị impressions lớn nhất:  3052003
</pre>
  
</details>

In [34]:
max_impression_index = np.where(impressions == max_impression)
max_impression_index

(array([525]),)

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>Vị trí impressions lớn nhất: 
</pre>
<pre>(array([525], dtype=int64),)</pre>
  
</details>

In [35]:
print('Spent tương ứng với impression lớn nhất:',spent[max_impression_index])

Spent tương ứng với impression lớn nhất: [639.95]


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>Giá trị spent tương ứng là: 
</pre>
  <pre>array([639.95])</pre>
</details>

In [37]:
# Câu 9: Tính tổng số click, impression và chi phí (spent) của tất cả các mẫu
print('Tổng số click:%d'%np.sum(clicks))
print('Tổng số impression:%d'%np.sum(impressions))
print('Tổng số spent:%.2f'%np.sum(spent))

Tổng số click:38165
Tổng số impression:213434828
Tổng số spent:58705.23


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>Tổng số click:  38165
Tổng số impression:  213434828
Tổng số chi phí:  58705.23
</pre>
  
</details>

In [44]:
# Câu 10: Cho biết click và impression của mẫu quảng cáo có mức spent lớn nhất
max_spent_index = np.where(spent == np.max(spent))
print('click của spent lớn nhất: %d'%clicks[max_spent_index])
print('impression của spent lớn nhất: %d'%impressions[max_spent_index])

click của spent lớn nhất: 340
impression của spent lớn nhất: 3052003


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>Giá trị spent lớn nhất:  639.95
click của mẫu quảng cáo có mức spent lớn nhất:  [1]
impression của mẫu quảng cáo có mức spent lớn nhất:  [7350]
</pre>
  
</details>

In [47]:
# Câu 11: Tìm giá trị trung bình và trung vị của clicks, impressions, spents
print('Trung bình của clicks: %.2f, trung vị của clicks: %.2f'%(np.mean(clicks),np.median(clicks)))
print('Trung bình của impressions: %.2f, trung vị của impressions: %.2f'%(np.mean(impressions),np.median(impressions)))
print('Trung bình của spent: %.2f, trung vị của spent: %.2f'%(np.mean(spent),np.median(spent)))

Trung bình của clicks: 33.39, trung vị của clicks: 8.00
Trung bình của impressions: 186732.13, trung vị của impressions: 51509.00
Trung bình của spent: 51.36, trung vị của spent: 12.37


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>Giá trị trung bình của clicks:  33.39020122484689
Giá trị trung vị của clicks:  8.0
Giá trị trung bình của impressions:  186732.1329833771
Giá trị trung vị của impressions:  51509.0
Giá trị trung bình của spents:  51.360656167979
Giá trị trung vị của spents:  12.37
</pre>
  
</details>

In [57]:
# Câu 12: Cho biết có những clicks nào >= 90 và <=100, tạo array clicks_above chứa các clicks này. 
clicks_above = np.extract((clicks >= 90) & (clicks <= 100), clicks)
clicks_above

array([ 95, 100,  91,  98,  91,  97,  90,  96,  97,  91, 100])

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>array([ 95, 100,  91,  98,  91,  97,  90,  96,  97,  91, 100])</pre>
  
</details>

In [58]:
# Có bao nhiêu clicks thỏa điều kiện?
clicks_above.size

11

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>11</pre>
  
</details>

In [59]:
# Câu 13: Tạo array spent_new lấy dữ liệu từ spent, 
# trong đó, các mẫu có spent > 147.8425 sẽ cập nhật lại bằng giá trị trung vị của spent.
spent_new = spent
spent_new[np.where(spent_new > 147.8425)] = np.median(spent_new)
spent_new

array([ 1.43,  1.82,  0.  , ..., 40.29, 12.37, 12.37])

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>array([ 1.43,  1.82,  0.  , ..., 40.29, 12.37, 12.37])</pre>
  
</details>

In [68]:
# Câu 14: Thêm 3 giá trị 5, 4.5, 10500 lần lượt vào cuối các array clicks, spents và impressions
clicks = np.append(clicks,[5])
spent = np.append(spent,[4.5])
impressions = np.append(impressions,[105000])
# Hiển thị 5 giá trị cuối của các array clicks, spent và impressions
print('5 giá trị cuối của clicks',clicks[-5:])
print('5 giá trị cuối của spent',spent[-5:])
print('5 giá trị cuối của impressions',impressions[-5:])

5 giá trị cuối của clicks [120  28 135 114   5]
5 giá trị cuối của spent [173.88  40.29 198.71 165.61   4.5 ]
5 giá trị cuối của impressions [637549. 151531. 790253. 513161. 105000.]


<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>array([120,  28, 135, 114,   5])</pre>
<pre>array([173.88,  40.29, 198.71, 165.61,   4.5 ])</pre>
<pre>array([637549, 151531, 790253, 513161,  10500])</pre>
  
</details>

In [69]:
# Câu 15: Xóa các giá trị ở index 10 và 100 của các array clicks, spents và impressions
clicks = np.delete(clicks, [10,100])
spent = np.delete(spent, [10,100])
impressions = np.delete(impressions, [10,100])

In [72]:
# Câu 16: Cho biết tương quan giữa click và impression
np.corrcoef(clicks, impressions)

array([[1.        , 0.94846678],
       [0.94846678, 1.        ]])

<details>
  <summary>Nhấn vào đây để xem kết quả!</summary>
  <pre>array([[1.        , 0.94849531],
       [0.94849531, 1.        ]])</pre>
  
</details>